### PR TITLE
Update controller code to use the controller-runtime helper function to create or update resources

### PR DESCRIPTION
**Pull request by jvanz**
## Description

This PR updates the controller code responsible to update the policy server resources to use the controller-runtime [`CreateOrUpdate`](https://pkg.go.dev/sigs.k8s.io/controller-runtime@v0.17.2/pkg/controller/controllerutil#CreateOrUpdate) helper function to abstract the logic of detecting if the resource should be created or updated. Making the code more simple and easy to maintain. 

During this process the unit tests testing the removed code have been removed as well. Which bring us to the spin off PR of this one, the  integration tests reorganization #731 .  The old tests have been reorganized to describe better the context of what the tests are validating. As well as updating the whole policy server controller tests to be independent as suggested by the Ginkgo documentation. After that, to ensure that the controller continue to works as before a bunch of integrations tests have been added covering the removed unit tests

~NOTE: I've not added test to cover the metrics configurations yet. I'll add a future PR to cover that. I do not want to wait more in this PR to get some feedback from the team.~  This is not necessary anymore. I've added tests for the metrics and tracing configuration here as well.

Fix https://github.com/kubewarden/kubewarden-controller/issues/723

## Test

```shell
make test
```

### Additional information

- This PR has the same content of the #726 . That PR is closed just to get a better history and code coverage report.

### CODE REVIEW

**Comment by jvanz** [comment]
@kubewarden/kubewarden-developers I take this opportunity to verify some of the policy controller test fail time to time. I could see the tests failing regularly in this branch. Therefore, I've started by reorganizing the tests to be independent. This way I can run the problematic test individually. That's why, you can see another test reorganization in the policy controller tests. ;)

**Comment by jvanz** [comment]
@kubewarden/kubewarden-developers I've split this PR into two. The second one, #731, has the test reorganization that I did while working on this change. As the tests added in this PR require the changes on #731, this PR is based on that changes. Thus, #731  should be merge before. 

Moving this PR to block until we merge #731.

File: internal/pkg/admission/policy-server-service_test.go [hunk 1]
```diff
@@ -1,28 +0,0 @@
...
-
-import (
-	"testing"
-
-	"github.com/kubewarden/kubewarden-controller/internal/pkg/constants"
-	policiesv1 "github.com/kubewarden/kubewarden-controller/pkg/apis/policies/v1"
-	"sigs.k8s.io/controller-runtime/pkg/client"
-)
-
-func TestMetricsEnabled(t *testing.T) {
```
Comment (viccuad) [inline internal/pkg/admission/policy-server-service_test.go:11]: I think we should keep the metrics test around.

Comment (jvanz) [reply internal/pkg/admission/policy-server-service_test.go:11]: Yes, I'm aware of this:

> NOTE: I've not added test to cover the metrics configurations yet. I'll add a future PR to cover that. I do not want to wait more in this PR to get some feedback from the team.

Comment (jvanz) [reply internal/pkg/admission/policy-server-service_test.go:11]: @viccuad I've added unit test to cover the metrics configuration in this PR. No need for future PR. No need to create convoluted integration tests.  It's easier to test this with unit test here. Please, take another look

**Review by viccuad — APPROVED** [review]
LGTM!

File: controllers/policyserver_controller_test.go [hunk 2]
```diff
@@ -142,59 +146,333 @@ var _ = Describe("PolicyServer controller", func() {
 		})
 	})
 
-	When("creating a new PolicyServer", func() {
-		var policyServer *policiesv1.PolicyServer
+	It("should use the policy server affinity configuration in the policy server deployment", func() {
```
Comment (fabriziosestito) [inline controllers/policyserver_controller_test.go:149]: can we group all the creation-related blocks into a `When("creating a PolicyServer")` ?

File: controllers/policyserver_controller_test.go [hunk 3]
```diff
@@ -142,59 +146,333 @@ var _ = Describe("PolicyServer controller", func() {
...
+						})),
+					})),
+				})),
+			})),
+		})))
+	})
 
-		BeforeEach(func() {
-			policyServer = policyServerFactory(policyServerName)
+	It("should creates policy server deployment with some default configuration", func() {
```
Comment (fabriziosestito) [inline controllers/policyserver_controller_test.go:186]: typo "create"

File: controllers/policyserver_controller_test.go [hunk 4]
```diff
@@ -142,59 +146,333 @@ var _ = Describe("PolicyServer controller", func() {
...
+			"FSGroup":             BeNil(),
+			"Sysctls":             BeNil(),
+			"FSGroupChangePolicy": BeNil(),
+			"SeccompProfile":      BeNil()})))
+
+		By("checking the deployment affinity")
+		Expect(deployment.Spec.Template.Spec.Affinity).To(BeNil())
+	})
+
+	It("should creates the policy server deployment and use the user defined security contexts", func() {
```
Comment (fabriziosestito) [inline controllers/policyserver_controller_test.go:226]: typo: "create"

File: controllers/policyserver_controller_test.go [hunk 5]
```diff
@@ -142,59 +146,333 @@ var _ = Describe("PolicyServer controller", func() {
...
+			"RunAsGroup":          BeNil(),
+			"RunAsNonRoot":        PointTo(BeFalse()),
+			"SupplementalGroups":  BeNil(),
+			"FSGroup":             BeNil(),
+			"Sysctls":             BeNil(),
+			"FSGroupChangePolicy": BeNil(),
+			"SeccompProfile":      BeNil()})))
+	})
+
+	It("should creates the policy server configmap empty if no policies are assigned ", func() {
```
Comment (fabriziosestito) [inline controllers/policyserver_controller_test.go:272]: typo: "create"

File: controllers/policyserver_controller_test.go [hunk 6]
```diff
@@ -142,59 +146,333 @@ var _ = Describe("PolicyServer controller", func() {
...
+		Expect(err).ToNot(HaveOccurred())
+		Expect(configmap).To(PointTo(MatchFields(IgnoreExtras, Fields{
+			"Data": MatchAllKeys(Keys{
+				constants.PolicyServerConfigPoliciesEntry: Equal("{}"),
+				constants.PolicyServerConfigSourcesEntry:  Equal("{}"),
+			}),
+		})))
+	})
+
+	It("should creates the policy server configmap with the assigned policies", func() {
```
Comment (fabriziosestito) [inline controllers/policyserver_controller_test.go:285]: typo: "create"

File: controllers/policyserver_controller_test.go [hunk 7]
```diff
@@ -142,59 +146,333 @@ var _ = Describe("PolicyServer controller", func() {
...
+		Expect(err).ToNot(HaveOccurred())
+		Expect(configMap).To(PointTo(MatchFields(IgnoreExtras, Fields{
+			"Data": MatchAllKeys(Keys{
+				constants.PolicyServerConfigPoliciesEntry: MatchJSON(policies),
+				constants.PolicyServerConfigSourcesEntry:  Equal("{}"),
+			}),
+		})))
+	})
+
+	It("should creates the policy server configmap with the sources authorities", func() {
```
Comment (fabriziosestito) [inline controllers/policyserver_controller_test.go:325]: typo: "create"

File: controllers/policyserver_controller_test.go [hunk 8]
```diff
@@ -142,59 +146,333 @@ var _ = Describe("PolicyServer controller", func() {
...
+				constants.PolicyServerConfigSourcesEntry:  MatchJSON(sources),
+			}),
+		})))
+	})
 
-		It("it should creates the policy server configmap empty when no policy is assigned", func() {
-			createPolicyServerAndWaitForItsService(policyServer)
-			Eventually(func() error {
-				_, err := getTestPolicyServerConfigMap(policyServerName)
+	It("should creates PodDisruptionBudget when policy server has MinAvailable configuration set", func() {
```
Comment (fabriziosestito) [inline controllers/policyserver_controller_test.go:363]: typo: "create"

File: controllers/policyserver_controller_test.go [hunk 9]
```diff
@@ -142,59 +146,333 @@ var _ = Describe("PolicyServer controller", func() {
...
+		policyServer.Spec.MinAvailable = &minAvailable
+		createPolicyServerAndWaitForItsService(policyServer)
+
+		Eventually(func() *k8spoliciesv1.PodDisruptionBudget {
+			pdb, _ := getPolicyServerPodDisruptionBudget(policyServerName)
+			return pdb
+		}, timeout, pollInterval).Should(policyServerPodDisruptionBudgetMatcher(policyServer, &minAvailable, nil))
+	})
+
+	It("should creates PodDisruptionBudget when policy server has MaxUnavailable configuration set", func() {
```
Comment (fabriziosestito) [inline controllers/policyserver_controller_test.go:376]: typo: "create"

File: controllers/policyserver_controller_test.go [hunk 10]
```diff
@@ -142,59 +146,333 @@ var _ = Describe("PolicyServer controller", func() {
...
+	It("should not create PodDisruptionBudget when policy server has no PDB configuration", func() {
+		policyServer := policyServerFactory(policyServerName)
+		createPolicyServerAndWaitForItsService(policyServer)
+		Consistently(func() error {
+			_, err := getPolicyServerPodDisruptionBudget(policyServerName)
+			return err
+		}, consistencyTimeout, pollInterval).ShouldNot(Succeed())
+	})
+
+	It("should creates the PolicyServer pod with the limits and the requests", func() {
```
Comment (fabriziosestito) [inline controllers/policyserver_controller_test.go:397]: typo: "create"

File: controllers/policyserver_controller_test.go [hunk 11]
```diff
@@ -142,59 +146,333 @@ var _ = Describe("PolicyServer controller", func() {
...
 
-		It("it should not create PDB when policy server is defined with no PodDisruptionBudget", func() {
-			createPolicyServerAndWaitForItsService(policyServer)
-			Consistently(func() error {
-				_, err := getPolicyServerPodDisruptionBudget(policyServerName)
+			return nil
+		}, timeout, pollInterval).Should(Succeed())
+	})
+
+	It("when the requests are updated should update the PolicyServer pod with the new requests", func() {
```
Comment (fabriziosestito) [inline controllers/policyserver_controller_test.go:430]: maybe rephrase (It)"should update the PolicyServer pod with the new requests when the requests are updated"

**Review by fabriziosestito — CHANGES_REQUESTED** [review]
Overall, LGTM.
Great work! 
The reconciler code looks very simple now and we got rid of all the custom logic.

Just left some minor comments.

File: internal/pkg/admission/policy-server-configmap.go [hunk 12]
```diff
@@ -200,7 +125,7 @@ func (policyConfigEntryMap PolicyConfigEntryMap) ToClusterAdmissionPolicyReconci
 	return res
 }
 
-func (r *Reconciler) createPoliciesMap(admissionPolicies []policiesv1.Policy) PolicyConfigEntryMap {
+func (r *Reconciler) CreatePoliciesMap(admissionPolicies []policiesv1.Policy) PolicyConfigEntryMap {
```
[the remark this answers is not in the document]
Comment (jvanz) [reply internal/pkg/admission/policy-server-configmap.go:128]: Sorry, I was using this function in the integration tests. But I've ended not using and generating the data to check the result there instead.

File: internal/pkg/admission/policy-server-configmap.go [hunk 13]
```diff
@@ -218,8 +143,8 @@ func (r *Reconciler) createPoliciesMap(admissionPolicies []policiesv1.Policy) Po
 	return policies
 }
 
-func (r *Reconciler) createSourcesMap(policyServer *policiesv1.PolicyServer) policyServerSourcesEntry {
-	sourcesEntry := policyServerSourcesEntry{}
+func (r *Reconciler) CreateSourcesMap(policyServer *policiesv1.PolicyServer) PolicyServerSourcesEntry {
```
[the remark this answers is not in the document]
Comment (jvanz) [reply internal/pkg/admission/policy-server-configmap.go:146]: Sorry, I was using this function in the integration tests. But I've ended not using and generating the data to check the result there instead.

**Comment by jvanz** [comment]
@fabriziosestito I've addressed your comments. Can you review it again? Thanks!

**Review by fabriziosestito — APPROVED** [review]

**Review by flavio — APPROVED** [review]
That's a really nice cleanup! :clap:
